### ISSUE TITLE
invoking AjaxErrorCallback<object> does not compile with 7.0.2

### ISSUE THREAD

**Issue by rekna1** [nikhilk/scriptsharp#22]
in some class, i have a function 

```
public void postData(string url, object data, AjaxRequestCallback<object> successCallback, AjaxErrorCallback<object>    errorCallback)
{
...
        if (!Script.IsNullOrUndefined(errorCallback))
              errorCallback(req, textStatus, error); // this line does compile with 7.0.0 but not with 7.0.2
}
```

The line which tries to invoke the errorCallback gives a compiler error:
_Check that your c# source compiles and that you are not using an unsupported feature._
It was compiling just fine with verison 7.0.0.

P.S. apparently the mscorlib and other dll's still have version number 7.0.0, while mscorlib.js has version 7.0.2. This is a little bit confusing, because until I looked at mscorlib.js I was thinking the scriptsharp.zip on your site was still containing setup for 7.0.0.
P.S.1 maybe it would be usefull to leave older version of scriptsharp downloadable on your site, so we can revert to an older version if necessary (I unfortunately didn't make a copy of the zip file, so I am left with an project which won't compile anymore)

**Comment by nikhilk** [comment]
I will post download links for previous versions tonight.

I didn't update the version #'s for the minor change because doing so requires everyone to change their csproj files (for the mscorlib reference), and wanted to save folks that trouble, but I should have updated the msi version... will keep in mind for the next update.

That said, I can't remember anything changing in 0.7.0.1 or 0.7.0.2 to affect this. I'll try it tonight and try to get a fix out, assuming I find the issue. I might need a bit more of the source for the repro (just guessing) ... can you include the relevant lines? I am guessing the line itself is within an anonymous delegate passed in into a jQuery method. Those details might be significant.

**Comment by rekna1** [comment]
This is more or less the complete function

```
    public void postData(string url, object data, AjaxRequestCallback<object> succesCallback, AjaxErrorCallback<object> errorCallback, string returnType, string requestType)
    {
        returnType = returnType ?? "text";
        requestType = requestType ?? "POST";

        jQuery.Ajax(new jQueryAjaxOptions(
            "cache", false,
            "data", data,
            "dataType", returnType,
            "error", (AjaxErrorCallback<object>)delegate(jQueryXmlHttpRequest<object> req, string textStatus, Exception error)
            {
                if (!Script.IsNullOrUndefined(errorCallback))
                    errorCallback(req, textStatus, error);
            },
            "success", (AjaxRequestCallback<object>)delegate(object dataSuccess, string textStatus, jQueryApi.jQueryXmlHttpRequest<object> request)
            {
                if (Script.IsNullOrUndefined(succesCallback))
                {
                    jQuery.Document.Append((string)dataSuccess);
                }
                else
                {
                    succesCallback(dataSuccess, textStatus, request);
                }
            },
            "type", requestType,
            "url", url
            ));
    }
```

**Comment by rekna1** [comment]
Even a very small test function, does not compile
            public void TestCallback(AjaxRequestCallback<object> test,jQueryXmlHttpRequest<object> req)
            {
                 test(null, "status", req);
             }

**Comment by nikhilk** [comment]
OK, I figured out the bug ... there is a subtle bug based on how generics are handled (aside: lots of little constraints, and limitations in that space, until the compiler fundamentally is changed to handle generics and other language features).

It has to do with the order of the way classes are laid out and it turns out jQueryAjaxOptions must have been read after jQueryErrorCallback in 0.7, but the c# compiler laid them out in reverse order in 0.7.0.2 for whatever reason.

I have a fix based on a bit of hack, and will try to clean it up for publishing.

However couple of things:
1. In your first example, you have a namespace qualified reference jQueryApi.jQueryXmlHttpRequest<object>. That needs to be just type name. Simple enough.
1. More limiting - in your second example, you have jQueryXmlHttpRequest, which is a generic type. You'll need to type it as jQueryXmlHttpRequest<object> or some other type for TData. You can't use jQueryXmlHttpRequest<TData> as the level of generics support allows you to use generic types, but not define new generic types/methods (in this class you won't be able to create your TestCallback as a generic method or declare the containing type as generic). Hopefully it won't be too limiting.

**Comment by rekna1** [comment]
Concerning my second example; something must have gone wrong with when pasting the code... I my code I am using the generic version but somehow the &lt;object&gt; was removed during pasting..  so it should have been:
        public void TestCallback( AjaxRequestCallback&lt;object&gt; test, jQueryXmlHttpRequest&lt;object&gt; req)

Hope fix will be posted soon, glad you found it!

**Comment by nikhilk** [comment]
Fixed in 0.7.1.

**Comment by nikhilk** [comment]
Unfortunately the problem wasn't fully fixed as shown in https://github.com/NikhilK/scriptsharp/issues/32.

Fixing it is going to require some breaking changes unfortunately - jQueryXmlHttpRequest<TData> will become jQueryDataHttpRequest and the jQueryXmlHttpRequest will be used for the non-generic name.

**Comment by nikhilk** [comment]
Fix included in 0.7.2.0 release.
